### DISCUSSION TITLE
The corners are too sharp

### DISCUSSION THREAD

**Discussion by joaopauloalbq** [yshui/picom#1021] [General]
> How is it now
> ![0 nor](https://user-images.githubusercontent.com/9905202/217555201-4b7e477b-fc80-4c1a-a674-925c7c7c8dd0.png)

> The end of the curve is rough.
> ![1 indi nor](https://user-images.githubusercontent.com/9905202/217555210-92ba5887-a7e6-4320-bd10-8a702c51b99b.png)

> Proposal: softer corners
> ![2 fix](https://user-images.githubusercontent.com/9905202/217555218-2c0affca-4ae3-4d83-898f-41e683047c49.png)

> How is it now (zoom)
![3 nor zoom](https://user-images.githubusercontent.com/9905202/217555224-704782ba-9b85-411b-9e3e-00ebbaaae6fc.png)

>  ->
> ![4 indi nor zoom](https://user-images.githubusercontent.com/9905202/217555227-7c754e11-afcb-4c9d-b460-a01b384da661.png)

> Softer corners (zoom)
![5 fix zoom](https://user-images.githubusercontent.com/9905202/217555229-a61c5f8b-a359-41a6-a869-592ffb97005e.png)

### picom version

v10.2
<details>
<summary>Diagnostics</summary>

**Version:** vgit-b700a

### Extensions:

* Shape: Yes
* XRandR: Yes
* Present: Present

### Misc:

* Use Overlay: No
  (Another compositor is already running)
* Config file used: /home/jp/.config/picom/picom.conf

### Drivers (inaccurate):

modesetting

### Backend: glx

* Driver vendors:
 * GLX: Mesa Project and SGI
 * GL: Intel
* GL renderer: Mesa Intel(R) UHD Graphics 620 (KBL GT2)
* Accelerated: 1

### Backend: egl

* Driver vendors:
 * EGL: Mesa Project
 * EGL driver: iris
 * GL: Intel
* GL renderer: Mesa Intel(R) UHD Graphics 620 (KBL GT2)
</details>

### Configuration:
<details>
    <summary>Configuration file</summary>

```
backend = "glx";
vsync = true;

glx-no-stencil = true;
glx-no-rebind-pixmap = true;
use-damage = false;

# Corners
corner-radius = 12
rounded-corners-exclude = [
  	"_NET_WM_STATE@:a *= '_NET_WM_STATE_MAXIMIZED_VERT'",
        "window_type = 'dock'",
        "window_type = 'dnd'",
  	"class_g = 'MEGAsync'"
];

# Shadow
shadow = true;
shadow-radius = 18;
shadow-offset-x = -18;
shadow-offset-y = -18;
shadow-opacity = 0.65;
shadow-exclude = [
  	"_NET_WM_WINDOW_TYPE:a *= '_KDE_NET_WM_WINDOW_TYPE_OVERRIDE'", # telegram
  	"_NET_WM_STATE@:a *= '_NET_WM_STATE_MAXIMIZED_VERT'",
  	"_NET_WM_STATE@:a *= '_NET_WM_STATE_FULLSCREEN'",
        "window_type = 'dnd'",
  	"class_g = 'MEGAsync'"
];
    
clip-shadow-above = [
  	"window_type = 'dock'"
]

detect-rounded-corners = true;

# Fading
fading = false;

focus-exclude = [
	# "class_g = 'slop'", # Maim
	"class_g = 'Rofi'"
];

mark-wmwin-focused = true;
```

**Comment by absolutelynothelix** [comment]
iirc, i've noticed this as well. i believe that roundness of corners may be improved for gl-powered backends (glx, egl), but they'll still be sharp on the xrender backend due to the x render extensions being not very flexible.

@tryone144, take a look, maybe?

**Reply by absolutelynothelix** [reply]
thinking about this i've tried to reimplement rounded corners using squircles (as suggested in #609). i believe that it looks better. however, it's just a quick and dirty prototype (can't get it working well because i'm bad at opengl).
![image](https://user-images.githubusercontent.com/52252283/221732289-8e1848cc-149c-4b14-a7ff-9e7c4fac432b.png)

**Reply by absolutelynothelix** [reply]
@yshui, @tryone144, what do you think about using squircles for rounded corners by default?

my work on this is available at https://github.com/absolutelynothelix/picom/commit/cac526a2a6f5a1b24f2bfe85361d784241afcfa0, but i can't make it friends with blur (and probably it's not well done because i'm bad at opengl and shaders), so i hope someone will pick it up if it'll be considered a good idea.
![image](https://user-images.githubusercontent.com/52252283/222001063-b8da5a94-94f2-43b6-bad4-698c9021f4ff.png)

**Reply by musjj** [reply]
Thank you for the patch, it still works on the latest commit. My corners looks beautiful now!
